### PR TITLE
compress-man-pages: skip compressed man pages

### LINKED ISSUE #21777: compress-man-pages.sh broken by recent changes in 'master'
The following derivation used to work fine:

~~~
    ledgerEnv = self.stdenv.mkDerivation {
      name = "ledger-env-0";
      unpackPhase = ":";
      installPhase = ''
        mkdir -p $out/{bin,share}
        ln -sv ${self.ledger2}/bin/ledger $out/bin/ledger2
        ln -sv ${self.ledger3}/bin/ledger $out/bin/ledger3
        ln -sv ledger3 $out/bin/ledger
        ln -sv ${self.ledger3}/share/man $out/share/
      '';
    };
~~~

But now it's broken in current Nixpkgs `master`, giving the following error message:

~~~~
[...]
'/nix/store/9mwh4azklkq62ss1rrpl0v0pv928q76f-ledger-env-0/share/man' -> '/nix/store/9szv7hrhplvw8b931bqv5fgpza91vcvb-ledger-3.1.1/share/man'
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/9mwh4azklkq62ss1rrpl0v0pv928q76f-ledger-env-0
gzipping man pages in /nix/store/9mwh4azklkq62ss1rrpl0v0pv928q76f-ledger-env-0
/nix/store/cc7k5l8k7q0lqsanciz3q4mili9bwxpy-compress-man-pages.sh: line 13: /nix/store/9mwh4azklkq62ss1rrpl0v0pv928q76f-ledger-env-0/share/man/man1/ledger.1.gz.gz: Permission denied
rm: cannot remove '/nix/store/9mwh4azklkq62ss1rrpl0v0pv928q76f-ledger-env-0/share/man/man1/ledger.1.gz.gz': No such file or directory
builder for ‘/nix/store/ayhif4mj1zmvx6hij1nc79baas791gji-ledger-env-0.drv’ failed with exit code 1
error: build of ‘/nix/store/ayhif4mj1zmvx6hij1nc79baas791gji-ledger-env-0.drv’ failed
~~~~

### PR DESCRIPTION
**Pull request by Mic92**
Because of bash 4.4 the semantics GLOBIGNORE changed.
This resulted in already compressed man pages to be compressed twice.

Work in progress, because I first need my number crunching machine to test this.
Currently it has only passed by eyeballs test.
fixes #21777

### CODE REVIEW

File: pkgs/build-support/setup-hooks/compress-man-pages.sh [hunk 1]
```diff
@@ -18,11 +21,16 @@ compressManPages() {
         fi
     done
 
+    # replace symlinks with uncompressed manpages
     for f in "$dir"/share/man/*/* "$dir"/share/man/*/*/*; do
-        if [ -L "$f" -a -f `readlink -f "$f"`.gz ]; then
-            ln -sf `readlink "$f"`.gz "$f".gz && rm "$f"
```
Comment (Mic92) [inline pkgs/build-support/setup-hooks/compress-man-pages.sh]: I wonder why the last `&&` is needed are build hooks executed with `set -e` anyway?

**Comment by vcunat** [comment]
Shellcheck is a nice tool:

> SC2166: Prefer [ p ] && [ q ] as [ p -a q ] is not well defined.

File: pkgs/build-support/setup-hooks/compress-man-pages.sh [hunk 2]
```diff
@@ -6,10 +6,13 @@ compressManPages() {
     if [ ! -d "$dir/share/man" ]; then return; fi
     echo "gzipping man pages in $dir"
 
-    GLOBIGNORE=.:..:*.gz:*.bz2
-
+    # compress all uncompressed manpages
     for f in "$dir"/share/man/*/* "$dir"/share/man/*/*/*; do
-        if [ -f "$f" -a ! -L "$f" ]; then
+        if [ "${f: -3}" != ".gz" \
```
Comment (copumpkin) [inline pkgs/build-support/setup-hooks/compress-man-pages.sh]: Why not use `=~` here? It feels awkward to have to mention the length of the right string to strip off that component on the left

Comment (Mic92) [inline pkgs/build-support/setup-hooks/compress-man-pages.sh]: done

File: pkgs/build-support/setup-hooks/compress-man-pages.sh [hunk 3]
```diff
@@ -6,10 +6,9 @@ compressManPages() {
     if [ ! -d "$dir/share/man" ]; then return; fi
     echo "gzipping man pages in $dir"
 
-    GLOBIGNORE=.:..:*.gz:*.bz2
-
+    # compress all uncompressed manpages
     for f in "$dir"/share/man/*/* "$dir"/share/man/*/*/*; do
-        if [ -f "$f" -a ! -L "$f" ]; then
+        if [ ! "$f" =~ .(bz2|gz)$ ]  && [ -f "$f" ] && [ ! -L "$f" ]; then
```
Comment (dezgeg) [inline pkgs/build-support/setup-hooks/compress-man-pages.sh]: `=~` needs to be used with the double bracket construct, IIRC.

Comment (vcunat) [inline pkgs/build-support/setup-hooks/compress-man-pages.sh]: Yes.
> SC2074: Can't use =~ in [ ]. Use [[..]] instead.

**Comment by vcunat** [comment]
Seems OK, except that the referenced "ledger" problem won't be gone, as `share/man/*/` may be traversing symlinks to wherever, and we also don't want that.

**Comment by vcunat** [comment]
I suppose it will be easiest to do this via `find`.  I'll test that.

**Comment by vcunat** [comment]
I'd go with 2064216dc80 (on top of this PR).  What do you think?

(I tested those ledger and docker cases; hopefully there's no other common case.)

**Comment by Mic92** [comment]
I agree on that.

**Pull request closed without merging**